File: frontend/lib/planner/balance.ts
import type { IndexedCatalog } from "@/lib/planner/catalog-types"
import type { PlanGraph, PlanNode } from "@/lib/planner/graph-types"

const POWER_TOLERANCE_MW = 0.01

export type NodeRates = {
  outputs: Record<string, number>
  inputs: Record<string, number>
  powerMW: number
}

export type EdgeBalance = {
  itemClass: string
  flowPerMin: number
  demandPerMin: number
  supplyPerMin: number
  overPerMin: number
  underPerMin: number
  recommendedMk: number
  capacityPerMin: number
  exceedsMax: boolean
  unit: string
}

export type MkRecommendation = {
  mk: number
  ratePerMin: number
  capacityPerMin: number
  exceedsMax: boolean
  unit: string
}

export type BalanceResult = {
  nodes: Record<string, NodeRates>
  edges: Record<string, EdgeBalance>
  totalPowerMW: number
}

export type PortStatus = {
  unterminatedOutputs: string[]
  starvedInputs: string[]
}

export function computePowerMW(
  powerBaseMW: number,
  clockPercent: number,
  powerExponent: number,
  somersloopCount: number,
  slotCount: number
): number {
  if (powerBaseMW <= 0) {
    return 0
  }
  const clockFactor = (clockPercent / 100) ** powerExponent
  let power = powerBaseMW * clockFactor
  if (slotCount > 0 && somersloopCount > 0) {
    const sloopFactor = (1 + somersloopCount / slotCount) ** 2
    power *= sloopFactor
  }
  return power
}

export function buildingPowerMW(
  catalog: IndexedCatalog,
  buildingClass: string,
  clockPercent: number,
  somersloopCount: number
): number | null {
  const building = catalog.buildingsByClass.get(buildingClass)
  if (!building) {
    return null
  }
  let sloops = somersloopCount
  if (!building.canChangeProductionBoost) {
    sloops = 0
  }
  return computePowerMW(
    building.powerBaseMW,
    clockPercent,
    building.powerExponent,
    sloops,
    building.somersloopSlots
  )
}

export function somersloopOutputMultiplier(
  building: IndexedCatalog["buildings"][number] | undefined,
  somersloopCount: number
): number {
  if (
    !building ||
    !building.canChangeProductionBoost ||
    building.somersloopSlots <= 0
  ) {
    return 1
  }
  if (somersloopCount <= 0) {
    return 1
  }
  return 1 + somersloopCount * building.somersloopBoostMultiplier
}

function nodeCount(count: number | undefined): number {
  if (!count || count <= 0) {
    return 1
  }
  return count
}

function beltMkNumber(className: string): number {
  if (className.includes("Mk6")) return 6
  if (className.includes("Mk5")) return 5
  if (className.includes("Mk4")) return 4
  if (className.includes("Mk3")) return 3
  if (className.includes("Mk2")) return 2
  return 1
}

function pipeMkNumber(className: string): number {
  if (className.includes("MK2") || className.includes("Mk2")) {
    return 2
  }
  return 1
}

export function recommendBeltMk(
  catalog: IndexedCatalog,
  ratePerMin: number
): MkRecommendation {
  const unit = "items/min"
  const sorted = [...catalog.belts].sort(
    (a, b) => beltMkNumber(a.className) - beltMkNumber(b.className)
  )
  if (sorted.length === 0) {
    return {
      mk: 0,
      ratePerMin,
      capacityPerMin: 0,
      exceedsMax: ratePerMin > 0,
      unit,
    }
  }
  const maxCap = sorted[sorted.length - 1].itemsPerMin
  for (const belt of sorted) {
    if (belt.itemsPerMin >= ratePerMin) {
      return {
        mk: beltMkNumber(belt.className),
        ratePerMin,
        capacityPerMin: belt.itemsPerMin,
        exceedsMax: false,
        unit,
      }
    }
  }
  return {
    mk: beltMkNumber(sorted[sorted.length - 1].className),
    ratePerMin,
    capacityPerMin: maxCap,
    exceedsMax: true,
    unit,
  }
}

export function recommendPipeMk(
  catalog: IndexedCatalog,
  ratePerMin: number
): MkRecommendation {
  const unit = "m³/min"
  const sorted = [...catalog.pipes].sort(
    (a, b) => pipeMkNumber(a.className) - pipeMkNumber(b.className)
  )
  if (sorted.length === 0) {
    return {
      mk: 0,
      ratePerMin,
      capacityPerMin: 0,
      exceedsMax: ratePerMin > 0,
      unit,
    }
  }
  const maxCap = sorted[sorted.length - 1].cubicMetersPerMin
  for (const pipe of sorted) {
    if (pipe.cubicMetersPerMin >= ratePerMin) {
      return {
        mk: pipeMkNumber(pipe.className),
        ratePerMin,
        capacityPerMin: pipe.cubicMetersPerMin,
        exceedsMax: false,
        unit,
      }
    }
  }
  return {
    mk: pipeMkNumber(sorted[sorted.length - 1].className),
    ratePerMin,
    capacityPerMin: maxCap,
    exceedsMax: true,
    unit,
  }
}

export function recommendTransportMk(
  catalog: IndexedCatalog,
  itemClass: string,
  ratePerMin: number
): MkRecommendation {
  const item = catalog.itemsByClass.get(itemClass)
  if (item && (item.form === "liquid" || item.form === "gas")) {
    return recommendPipeMk(catalog, ratePerMin)
  }
  return recommendBeltMk(catalog, ratePerMin)
}

export function analyzeGraph(
  catalog: IndexedCatalog,
  graph: PlanGraph
): BalanceResult {
  const nodeRates: Record<string, NodeRates> = {}
  let totalPower = 0

  for (const node of graph.nodes) {
    const rates: NodeRates = { outputs: {}, inputs: {}, powerMW: 0 }

    switch (node.role) {
      case "source": {
        if (node.itemClass && (node.ratePerMin ?? 0) > 0) {
          rates.outputs[node.itemClass] =
            (node.ratePerMin ?? 0) * nodeCount(node.count)
        }
        break
      }
      case "sink": {
        if (node.itemClass) {
          rates.inputs[node.itemClass] = node.ratePerMin ?? 0
        }
        break
      }
      case "process":
      default: {
        if (node.recipeClass && node.buildingClass) {
          const recipe = catalog.recipesByClass.get(node.recipeClass)
          const building = catalog.buildingsByClass.get(node.buildingClass)
          if (recipe && building) {
            const mult = nodeCount(node.count)
            const clock = node.clockPercent && node.clockPercent > 0
              ? node.clockPercent
              : 100
            const sloopMult = somersloopOutputMultiplier(
              building,
              node.somersloopCount ?? 0
            )
            const cyclesPerMin =
              (60 / recipe.durationSec) *
              building.manufacturingSpeed *
              (clock / 100) *
              mult *
              sloopMult
            for (const product of recipe.products) {
              rates.outputs[product.className] = cyclesPerMin * product.amount
            }
            for (const ingredient of recipe.ingredients) {
              rates.inputs[ingredient.className] =
                cyclesPerMin * ingredient.amount
            }
            const power = buildingPowerMW(
              catalog,
              node.buildingClass,
              clock,
              node.somersloopCount ?? 0
            )
            if (power !== null) {
              rates.powerMW = power * mult
              totalPower += rates.powerMW
            }
          }
        }
        break
      }
    }

    nodeRates[node.id] = rates
  }

  const edgeFlows: Record<string, EdgeBalance> = {}

  for (const edge of graph.edges) {
    const src = nodeRates[edge.sourceNodeId] ?? {
      outputs: {},
      inputs: {},
      powerMW: 0,
    }
    const dst = nodeRates[edge.targetNodeId] ?? {
      outputs: {},
      inputs: {},
      powerMW: 0,
    }
    const supply = src.outputs[edge.itemClass] ?? 0
    const demand = dst.inputs[edge.itemClass] ?? 0

    let sameItemOutgoing = 0
    for (const other of graph.edges) {
      if (
        other.sourceNodeId === edge.sourceNodeId &&
        other.itemClass === edge.itemClass
      ) {
        sameItemOutgoing++
      }
    }

    const available =
      sameItemOutgoing > 0 ? supply / sameItemOutgoing : 0
    let flow = available
    if (demand > 0 && flow > demand) {
      flow = demand
    }

    const rec = recommendTransportMk(catalog, edge.itemClass, flow)

    edgeFlows[edge.id] = {
      itemClass: edge.itemClass,
      flowPerMin: flow,
      demandPerMin: demand,
      supplyPerMin: supply,
      overPerMin: Math.max(0, available - demand),
      underPerMin: Math.max(0, demand - flow),
      recommendedMk: rec.mk,
      capacityPerMin: rec.capacityPerMin,
      exceedsMax: rec.exceedsMax,
      unit: rec.unit,
    }
  }

  return {
    nodes: nodeRates,
    edges: edgeFlows,
    totalPowerMW: totalPower,
  }
}

export function analyzePortStatus(
  catalog: IndexedCatalog,
  graph: PlanGraph,
  node: PlanNode
): PortStatus {
  const unterminatedOutputs: string[] = []
  const starvedInputs: string[] = []

  const connectedOut = new Set<string>()
  const connectedIn = new Set<string>()

  for (const edge of graph.edges) {
    if (edge.sourceNodeId === node.id) {
      connectedOut.add(edge.sourcePort)
    }
    if (edge.targetNodeId === node.id) {
      connectedIn.add(edge.targetPort)
    }
  }

  if (node.role === "source" && node.itemClass) {
    const port = `out:${node.itemClass}`
    if (!connectedOut.has(port)) {
      unterminatedOutputs.push(node.itemClass)
    }
    return { unterminatedOutputs, starvedInputs }
  }

  if (node.role === "sink" && node.itemClass) {
    const port = `in:${node.itemClass}`
    if (!connectedIn.has(port)) {
      starvedInputs.push(node.itemClass)
    }
    return { unterminatedOutputs, starvedInputs }
  }

  if (node.role === "process" && node.recipeClass) {
    const recipe = catalog.recipesByClass.get(node.recipeClass)
    if (recipe) {
      for (const product of recipe.products) {
        const port = `out:${product.className}`
        if (!connectedOut.has(port)) {
          unterminatedOutputs.push(product.className)
        }
      }
      for (const ingredient of recipe.ingredients) {
        const port = `in:${ingredient.className}`
        if (!connectedIn.has(port)) {
          starvedInputs.push(ingredient.className)
        }
      }
    }
  }

  return { unterminatedOutputs, starvedInputs }
}

export function withinPowerTolerance(got: number, want: number): boolean {
  const got2 = Math.round(got * 100) / 100
  const want2 = Math.round(want * 100) / 100
  if (Math.abs(got2 - want2) <= POWER_TOLERANCE_MW) {
    return true
  }
  return Math.abs(Math.round(got * 10) / 10 - want) <= POWER_TOLERANCE_MW
}

File: backend/internal/planner/balance.go
package planner

import (
	"math"
	"sort"
	"strings"
)

const powerToleranceMW = 0.01

// NodeRates holds per-item production/consumption rates for one graph node.
type NodeRates struct {
	Outputs map[string]float64 `json:"outputs"`
	Inputs  map[string]float64 `json:"inputs"`
	PowerMW float64            `json:"powerMW"`
}

// EdgeBalance is derived flow data for one connection.
type EdgeBalance struct {
	ItemClass       string  `json:"itemClass"`
	FlowPerMin      float64 `json:"flowPerMin"`
	DemandPerMin    float64 `json:"demandPerMin"`
	SupplyPerMin    float64 `json:"supplyPerMin"`
	OverPerMin      float64 `json:"overPerMin"`
	UnderPerMin     float64 `json:"underPerMin"`
	RecommendedMk   int     `json:"recommendedMk"`
	CapacityPerMin  float64 `json:"capacityPerMin"`
	ExceedsMax      bool    `json:"exceedsMax"`
	Unit            string  `json:"unit"`
}

// MkRecommendation is belt/pipe Mk guidance for a numeric rate.
type MkRecommendation struct {
	Mk             int     `json:"mk"`
	RatePerMin     float64 `json:"ratePerMin"`
	CapacityPerMin float64 `json:"capacityPerMin"`
	ExceedsMax     bool    `json:"exceedsMax"`
	Unit           string  `json:"unit"`
}

// GraphNode is the persisted process/source node shape used by balance.
type GraphNode struct {
	ID              string  `json:"id"`
	Role            string  `json:"role"`
	RecipeClass     string  `json:"recipeClass"`
	BuildingClass   string  `json:"buildingClass"`
	ItemClass       string  `json:"itemClass"`
	Count           float64 `json:"count"`
	ClockPercent    float64 `json:"clockPercent"`
	SomersloopCount int     `json:"somersloopCount"`
	RatePerMin      float64 `json:"ratePerMin"`
}

// GraphEdge connects two node ports.
type GraphEdge struct {
	ID            string `json:"id"`
	SourceNodeID  string `json:"sourceNodeId"`
	SourcePort    string `json:"sourcePort"`
	TargetNodeID  string `json:"targetNodeId"`
	TargetPort    string `json:"targetPort"`
	ItemClass     string `json:"itemClass"`
}

// Graph is the minimal graph_json shape for balance analysis.
type Graph struct {
	Nodes []GraphNode `json:"nodes"`
	Edges []GraphEdge `json:"edges"`
}

// BalanceResult is the derived balance output for a graph.
type BalanceResult struct {
	Nodes map[string]NodeRates          `json:"nodes"`
	Edges map[string]EdgeBalance        `json:"edges"`
	TotalPowerMW float64                `json:"totalPowerMW"`
}

// ComputePowerMW returns building power draw using wiki formula from proposal §4.1.
func ComputePowerMW(powerBaseMW, clockPercent float64, powerExponent float64, somersloopCount, slotCount int) float64 {
	if powerBaseMW <= 0 {
		return 0
	}
	clockFactor := math.Pow(clockPercent/100, powerExponent)
	power := powerBaseMW * clockFactor
	if slotCount > 0 && somersloopCount > 0 {
		sloopFactor := math.Pow(1+float64(somersloopCount)/float64(slotCount), 2)
		power *= sloopFactor
	}
	return power
}

// BuildingPowerMW computes power for a catalog building with overrides.
func BuildingPowerMW(cat *Catalog, buildingClass string, clockPercent float64, somersloopCount int) (float64, bool) {
	building, ok := cat.BuildingByClass(buildingClass)
	if !ok {
		return 0, false
	}
	slots := building.SomersloopSlots
	if !building.CanChangeProductionBoost {
		somersloopCount = 0
	}
	return ComputePowerMW(building.PowerBaseMW, clockPercent, building.PowerExponent, somersloopCount, slots), true
}

// SomersloopOutputMultiplier returns the production boost from filled sloops.
func SomersloopOutputMultiplier(building *Building, somersloopCount int) float64 {
	if building == nil || !building.CanChangeProductionBoost || building.SomersloopSlots <= 0 {
		return 1
	}
	if somersloopCount <= 0 {
		return 1
	}
	return 1 + float64(somersloopCount)*building.SomersloopBoostMultiplier
}

// RecipeOutputPerMinute returns the per-building output rate for one product.
func RecipeOutputPerMinute(cat *Catalog, recipeClass, productClass, buildingClass string, clockPercent float64, somersloopCount int) (float64, bool) {
	recipe, ok := cat.recipesByClass[recipeClass]
	if !ok {
		return 0, false
	}
	building, ok := cat.BuildingByClass(buildingClass)
	if !ok {
		return 0, false
	}

	var amount float64
	for _, product := range recipe.Products {
		if product.ClassName == productClass {
			amount = product.Amount
			break
		}
	}
	if amount <= 0 {
		return 0, false
	}

	cyclesPerMin := (60 / recipe.DurationSec) * building.ManufacturingSpeed * (clockPercent / 100)
	rate := cyclesPerMin * amount * SomersloopOutputMultiplier(building, somersloopCount)
	return rate, true
}

// RecommendBeltMk returns the smallest belt Mk whose capacity covers rate items/min.
func (c *Catalog) RecommendBeltMk(ratePerMin float64) MkRecommendation {
	unit := "items/min"
	if len(c.Belts) == 0 {
		return MkRecommendation{RatePerMin: ratePerMin, Unit: unit, ExceedsMax: ratePerMin > 0}
	}

	sorted := append([]Belt(nil), c.Belts...)
	sort.Slice(sorted, func(i, j int) bool {
		return beltMkNumber(sorted[i].ClassName) < beltMkNumber(sorted[j].ClassName)
	})

	maxCap := sorted[len(sorted)-1].ItemsPerMin
	for _, belt := range sorted {
		if belt.ItemsPerMin >= ratePerMin {
			return MkRecommendation{
				Mk:             beltMkNumber(belt.ClassName),
				RatePerMin:     ratePerMin,
				CapacityPerMin: belt.ItemsPerMin,
				ExceedsMax:     false,
				Unit:           unit,
			}
		}
	}
	return MkRecommendation{
		Mk:             beltMkNumber(sorted[len(sorted)-1].ClassName),
		RatePerMin:     ratePerMin,
		CapacityPerMin: maxCap,
		ExceedsMax:     true,
		Unit:           unit,
	}
}

// RecommendPipeMk returns the smallest pipe Mk whose capacity covers rate m³/min.
func (c *Catalog) RecommendPipeMk(ratePerMin float64) MkRecommendation {
	unit := "m³/min"
	if len(c.Pipes) == 0 {
		return MkRecommendation{RatePerMin: ratePerMin, Unit: unit, ExceedsMax: ratePerMin > 0}
	}

	sorted := append([]Pipe(nil), c.Pipes...)
	sort.Slice(sorted, func(i, j int) bool {
		return pipeMkNumber(sorted[i].ClassName) < pipeMkNumber(sorted[j].ClassName)
	})

	maxCap := sorted[len(sorted)-1].CubicMetersPerMin
	for _, pipe := range sorted {
		if pipe.CubicMetersPerMin >= ratePerMin {
			return MkRecommendation{
				Mk:             pipeMkNumber(pipe.ClassName),
				RatePerMin:     ratePerMin,
				CapacityPerMin: pipe.CubicMetersPerMin,
				ExceedsMax:     false,
				Unit:           unit,
			}
		}
	}
	return MkRecommendation{
		Mk:             pipeMkNumber(sorted[len(sorted)-1].ClassName),
		RatePerMin:     ratePerMin,
		CapacityPerMin: maxCap,
		ExceedsMax:     true,
		Unit:           unit,
	}
}

// RecommendTransportMk picks belt or pipe guidance based on item form.
func (c *Catalog) RecommendTransportMk(itemClass string, ratePerMin float64) MkRecommendation {
	if item, ok := c.ItemByClass(itemClass); ok && (item.Form == ItemFormLiquid || item.Form == ItemFormGas) {
		return c.RecommendPipeMk(ratePerMin)
	}
	return c.RecommendBeltMk(ratePerMin)
}

// AnalyzeGraph computes node rates, edge flows, imbalance, and total power.
func (c *Catalog) AnalyzeGraph(graph Graph) BalanceResult {
	nodeRates := make(map[string]NodeRates, len(graph.Nodes))
	var totalPower float64

	for _, node := range graph.Nodes {
		rates := NodeRates{
			Outputs: map[string]float64{},
			Inputs:  map[string]float64{},
		}

		switch node.Role {
		case "source":
			if node.ItemClass != "" && node.RatePerMin > 0 {
				rates.Outputs[node.ItemClass] = node.RatePerMin * nodeCount(node.Count)
			}
		case "process", "":
			if node.RecipeClass != "" && node.BuildingClass != "" {
				recipe, ok := c.recipesByClass[node.RecipeClass]
				if ok {
					building, ok := c.BuildingByClass(node.BuildingClass)
					if ok {
						mult := nodeCount(node.Count)
						clock := node.ClockPercent
						if clock <= 0 {
							clock = 100
						}
						sloopMult := SomersloopOutputMultiplier(building, node.SomersloopCount)
						cyclesPerMin := (60 / recipe.DurationSec) * building.ManufacturingSpeed * (clock / 100) * mult * sloopMult
						for _, product := range recipe.Products {
							rates.Outputs[product.ClassName] = cyclesPerMin * product.Amount
						}
						for _, ingredient := range recipe.Ingredients {
							rates.Inputs[ingredient.ClassName] = cyclesPerMin * ingredient.Amount
						}
						power, _ := BuildingPowerMW(c, node.BuildingClass, clock, node.SomersloopCount)
						power *= mult
						rates.PowerMW = power
						totalPower += power
					}
				}
			}
		}
		nodeRates[node.ID] = rates
	}

	edgeFlows := make(map[string]EdgeBalance, len(graph.Edges))

	for _, edge := range graph.Edges {
		src := nodeRates[edge.SourceNodeID]
		dst := nodeRates[edge.TargetNodeID]
		supply := src.Outputs[edge.ItemClass]
		demand := dst.Inputs[edge.ItemClass]

		sameItemOutgoing := 0
		for _, other := range graph.Edges {
			if other.SourceNodeID == edge.SourceNodeID && other.ItemClass == edge.ItemClass {
				sameItemOutgoing++
			}
		}
		available := 0.0
		if sameItemOutgoing > 0 {
			available = supply / float64(sameItemOutgoing)
		}
		flow := available
		if demand > 0 && flow > demand {
			flow = demand
		}
		rec := c.RecommendTransportMk(edge.ItemClass, flow)

		edgeFlows[edge.ID] = EdgeBalance{
			ItemClass:      edge.ItemClass,
			FlowPerMin:     flow,
			DemandPerMin:   demand,
			SupplyPerMin:   supply,
			OverPerMin:     math.Max(0, available-demand),
			UnderPerMin:    math.Max(0, demand-flow),
			RecommendedMk:  rec.Mk,
			CapacityPerMin: rec.CapacityPerMin,
			ExceedsMax:     rec.ExceedsMax,
			Unit:           rec.Unit,
		}
	}

	return BalanceResult{
		Nodes:        nodeRates,
		Edges:        edgeFlows,
		TotalPowerMW: totalPower,
	}
}

func beltMkNumber(className string) int {
	switch {
	case strings.Contains(className, "Mk6"):
		return 6
	case strings.Contains(className, "Mk5"):
		return 5
	case strings.Contains(className, "Mk4"):
		return 4
	case strings.Contains(className, "Mk3"):
		return 3
	case strings.Contains(className, "Mk2"):
		return 2
	default:
		return 1
	}
}

func pipeMkNumber(className string) int {
	if strings.Contains(className, "MK2") || strings.Contains(className, "Mk2") {
		return 2
	}
	return 1
}

// WithinPowerTolerance reports whether two MW values match within wiki test tolerance.
func WithinPowerTolerance(got, want float64) bool {
	got2 := math.Round(got*100) / 100
	want2 := math.Round(want*100) / 100
	if math.Abs(got2-want2) <= powerToleranceMW {
		return true
	}
	// Wiki tables sometimes publish one-decimal MW (e.g. 53.7 for amplified 250% constructor).
	return math.Abs(math.Round(got*10)/10-want) <= powerToleranceMW
}

func nodeCount(count float64) float64 {
	if count <= 0 {
		return 1
	}
	return count
}
